share/shwap/p2p/shrex/shrexnd/server.go

```diff
@@ -0,0 +1,219 @@
+package shrexnd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	"go.uber.org/zap"
+
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	"github.com/celestiaorg/celestia-node/libs/utils"
+	eds "github.com/celestiaorg/celestia-node/share/new_eds"
+	"github.com/celestiaorg/celestia-node/share/shwap"
+	"github.com/celestiaorg/celestia-node/share/shwap/p2p/shrex"
+	shrexpb "github.com/celestiaorg/celestia-node/share/shwap/p2p/shrex/pb"
+	"github.com/celestiaorg/celestia-node/store"
+)
+
+// Server implements server side of shrex/nd protocol to serve namespaced share to remote
+// peers.
+type Server struct {
+	cancel context.CancelFunc
+
+	host       host.Host
+	protocolID protocol.ID
+
+	handler network.StreamHandler
+	store   *store.Store
+
+	params     *Parameters
+	middleware *shrex.Middleware
+	metrics    *shrex.Metrics
+}
+
+// NewServer creates new Server
+func NewServer(params *Parameters, host host.Host, store *store.Store) (*Server, error) {
+	if err := params.Validate(); err != nil {
+		return nil, fmt.Errorf("shrex-nd: server creation failed: %w", err)
+	}
+
+	srv := &Server{
+		store:      store,
+		host:       host,
+		params:     params,
+		protocolID: shrex.ProtocolID(params.NetworkID(), protocolString),
+		middleware: shrex.NewMiddleware(params.ConcurrencyLimit),
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	srv.cancel = cancel
```

Comment:
It should be part of Start as we usually do. I know it's not thread-safe, but it's safe to assume the constructor, Start, and Stop are called sequentially. It's confusing when we have it differently.

Reply:
also part of the start in eds server